### PR TITLE
Add `haven` option to `raw_or_label`

### PR DESCRIPTION
**Pull request by ezraporter**
# Description
This PR modifies `read_redcap()` to let users specify `raw_or_label = "haven"` and have categorical fields converted to `haven_labelled` vectors instead of factors. I kept `labelled` in `Suggests` and added a check that the user has it installed if they specify `raw_or_label = "haven"`.

# Implementation considerations

`haven_labelled` vectors, unlike factors, preserve the underlying data values/types. The non-trivial part about this is that to apply the labels, the data types in the data must match the data types of the vector of values we read from the metadata. This is slightly tricky because there's no foolproof way to say "cast vector x to type of vector y" for arbitrary y. My approach was to basically do our best at casting using `readr`'s parsing functions and fall back to converting the underlying values to `chr` if all else fails.

See below for a concrete example if it's helpful.

# Proposed Changes 
- Add `"haven"` option to `read_redcap()` `raw_or_label` and update appropriate arg checks
- Refactor `multi_choice_to_labels()`
    - Now takes a `raw_or_label` argument from `read_redcap()`
    - Passes off logic for how categorical fields are handled to a `label_handler` function which will be one of `apply_labs_factor()` or `apply_labs_haven()`

# Worked Example

Suppose we have a redcap field coded like this:
|value|label|
|-|-|
|3|apple|
|5|orange|
|9|banana|

The `db_data` we get from redcap will have a field like this:

```r
 my_field
    <int>
        5
        3
        5
        9
```

where the `int` datatype was actually determined by `readr` since we don't allow users to pass data type specifications to `redcap_read_oneshot()`

The `db_metadata` will contain

```r
 field_name   select_choices_or_calculations         
 <chr>        <chr>                                  
 my_field     3, apple | 5, orange | 9, banana
```

where the values are stored as `chr` rather than `int`.

My implementation:
1. Checks the data type of `db_data$my_field` and finds it's `int`
2. Casts the labels read from `db_metadata`, `c(apple = "3", orange = "5", banana = "9")`, to `int` with `readr::parse_integer()`
3. Applies the labels with `labelled::set_value_labels()`

If step 2 had failed for some reason then it would cast `db_data$my_field` to `chr` and apply the labels to that.

### Issue Addressed
Related to #178

### PR Checklist
Before submitting this PR, please check and verify below that the submission meets the below criteria:

- [x] New/revised functions have associated tests
- [x] New/revised functions that update downstream outputs have associated static testing files (`.RDS`) updated under `inst/testdata/create_test_data.R`
- [x] New/revised functions use appropriate naming conventions
- [x] New/revised functions don't repeat code
- [x] Code changes are less than **250** lines total
- [x] Issues linked to the PR using [GitHub's list of keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] The appropriate reviewer is assigned to the PR
- [x] The appropriate developers are assigned to the PR
- [x] Pre-release package version incremented using `usethis::use_version()`

# Code Review
This section to be used by the reviewer and developers during Code Review after PR submission

### Code Review Checklist

- [x] I checked that new files follow naming conventions and are in the right place
- [x] I checked that documentation is complete, clear, and without typos
- [x] I added/edited comments to explain "why" not "how"
- [x] I checked that all new variable and function names follow naming conventions
- [x] I checked that new tests have been written for key business logic and/or bugs that this PR fixes
- [x] I checked that new tests address important edge cases

### CODE REVIEW

File: R/utils.R [hunk 1]
```diff
@@ -485,15 +472,122 @@ multi_choice_to_labels <- function(db_data, db_metadata) {
...
+force_cast <- function(x, ptype) {
+  ptype <- vec_ptype(ptype)
+  if (is.logical(ptype)) {
+    out <- parse_logical(x)
+  } else if (is.integer(ptype)) {
+    out <- parse_integer(x)
+  } else if (is.numeric(ptype)) {
+    out <- parse_double(x)
+  } else if (is.Date(ptype)) {
+    out <- parse_date(x)
```
Comment (rsh52) [inline R/utils.R:578]: I think we can address these Codecov warnings by adding the associated data types to tests in your new tests. I don't think it's useful to write a test for this function in particular since it's so basic, but you could do either.

https://github.com/CHOP-CGTInformatics/REDCapTidieR/blob/c58e78363fa246589a0bbfd1d05554c1164f9e01/tests/testthat/test-utils.R#L436

File: R/utils.R [hunk 2]
```diff
@@ -485,15 +472,122 @@ multi_choice_to_labels <- function(db_data, db_metadata) {
...
   }
   db_data
 }
 
+#' @title
+#' Apply factor labels to a vector
+#'
+#' @param x a vector to label
+#' @param labels a named vector of labels in the format `c(value = label)`
+#' @param \dots unused
```
Comment (rsh52) [inline R/utils.R:491]: Do we even need `...` as an option then? Is this just for future-proofing?

File: R/utils.R [hunk 3]
```diff
@@ -485,15 +472,122 @@ multi_choice_to_labels <- function(db_data, db_metadata) {
...
+#' Assumes a check_installed() has been run for `labelled`. Since `haven` preserves the
+#' underlying data values we need to make sure the data type of the value options in the metadata matches
+#' the data type of the values in the actual data. This function accepts a prototype, usually a column
+#' from db_data, and uses `force_cast()` to do a best-effort casting of the value options in the metadata
+#' to the same data type as `ptype`. The fallback is to convert `x` and the value labels to character.
+#'
+#' @param x a vector to label
+#' @param labels a named vector of labels in the format `c(value = label)`
+#' @param ptype vector to serve as prototype for label values
+#' @param \dots unused
```
Comment (rsh52) [inline R/utils.R:516]: Ditto above

File: R/utils.R [hunk 4]
```diff
@@ -485,15 +472,122 @@ multi_choice_to_labels <- function(db_data, db_metadata) {
...
+  labels_cast <- withCallingHandlers(
+    {
+      force_cast(labels, ptype)
+    },
+    warning = function(w) {
+      cnd <<- w
+      cnd_muffle(w)
+    }
+  )
+  if (!is.null(attr(labels_cast, "problems"))) {
```
Comment (rsh52) [inline R/utils.R:537]: Just a thought, should this be transitioned to a new `check_*()` function for organizational purposes?

**Review by rsh52 — APPROVED** [review]
Great work! Some minor comments, but I think the output looks like what's needed and there were some clever methods used here. 

I would next see if we can get the issuer to test this and get their feedback on if it meets expectations.

File: R/utils.R (hunk 2 above)
Comment (ezraporter) [reply R/utils.R:491]: In this case we do need it because this function gets called as `label_handler` which takes a `ptype` when `label_handler <- apply_labs_haven`. The `...` just allows that `ptype` to get passed through unused. I'll make the annotation clearer.

File: R/utils.R (hunk 3 above)
Comment (ezraporter) [reply R/utils.R:516]: In this case we actually don't need it since the parameters of `apply_labs_haven()` are a superset of `apply_labs_factor()`. I kept it around to mirror `apply_labs_factor` but we can remove

File: R/utils.R (hunk 4 above)
Comment (ezraporter) [reply R/utils.R:537]: I'm not opposed to this. My thought was that these checks are more logic that implements what `apply_labs_haven()` does rather than general data/arg checks like we have in the `check_*()` stuff. I worry that it will actually be less clear to separate this from the `withCallingHandlers()` above.

File: R/utils.R (hunk 1 above)
Comment (ezraporter) [reply R/utils.R:578]: Good call. Will do! Btw are those code coverage warnings getting added to the conversation new? I've never seen them before

Comment (rsh52) [reply R/utils.R:578]: Unsure, I started noticing them in the past month or two. Relatively helpful for helping keep our % up

File: R/utils.R (hunk 4 above)
Comment (rsh52) [reply R/utils.R:537]: Fair enough, leave as is then

File: R/utils.R (hunk 1 above)
Comment (ezraporter) [reply R/utils.R:582]: Okay interesting finding (and I think why I skipped including those tests to start).

With `labelled` if you try to apply labels to vector that has no corresponding `val_labels<-` method it just returns the unlabelled vector:

```r
labels <- as.Date(c("The first day of 2023" = "2023-01-01"))

x <- labelled::set_value_labels(as.Date("2023-01-01"), .labels = labels)

x
#> [1] "2023-01-01:

class(x)
#> [1] "Date"
```

That means if someone has some crazy redcap where the value options are dates and they do `raw_or_label = "haven"`, their value labels won't get applied at all. BUT if they went ahead and implemented a `val_labels<-.Date` method then they would get applied. This is such a weird situation I don't think we should even care tbh. I added tests for `force_cast()` just to get rid of the coverage warnings.

Comment (rsh52) [reply R/utils.R:582]: Huh, very interesting find. In that case yea, much better to just test `force_cast()`. If for _some_ reason someone encounters this then an issue can be opened. 

I figured we wouldn't test `force_cast()` since most of the options for it were already tested in the parent function but whatever appeases the CodeCov gods is fine.

Comment (larmarange) [reply R/utils.R:582]: Value labels could be applied only to character or numeric vectors. It is not possible to add value labels to date vectors.

In fact, an error should have occured. I have open an issue on labelled website: https://github.com/larmarange/labelled/issues/156
